src/workerd/api/global-scope.c++

```diff
@@ -508,27 +527,33 @@ TimeoutId::NumberType ServiceWorkerGlobalScope::setTimeoutInternal(
 }
 
 TimeoutId::NumberType ServiceWorkerGlobalScope::setTimeout(
+    jsg::Lock& js,
     jsg::V8Ref<v8::Function> function,
     jsg::Optional<double> msDelay,
-    jsg::Varargs args,
-    v8::Isolate* isolate) {
+    jsg::Varargs args) {
+  if (js.isAsyncContextTrackingEnabled()) {
+    // If isAsyncContextTrackingEnabled returns false, then we assume that the root context
+    // is being used and we don't actually have to wrap the function.
```

Comment:
Maybe worth noting that the reason this assumption holds is because setTimeout() callbacks are always invoked by the system, hence without any `AsyncContextFrame::Scope`s on the stack. Hence even if async context tracking is enabled later, the callback will still be invoked using the root frame.

And on that note, maybe the check should not be for `js.isAsyncContextTrackingEnabled()`, but rather checking whether the current frame is the root. This way it's clearer that the logic doesn't change if async context tracking is enabled later on, and also the optimization still applies even when async context tracking is enabled.

On a related note, I kind of thing we should remove the `js.isAsyncContextTrackingEnabled()` method entirely, because it seems easy to use wrong -- if you forget to consider that async context tracking may kick in lazily.